### ISSUE TITLE
Running infection with PHPUnit result cache in subdirectories fails

### ISSUE THREAD

**Issue by localheinz** [infection/infection#572]
| Question    | Answer
| ------------| ---------------
| Infection version | `0.11.1`
| Test Framework version | `phpunit/phpunit:7.4.3`
| PHP version | `7.1.22`
| Platform    | macOs
| Github Repo | n/a |

Related to #487.

Running `infection` with a `phpunit` configuration where result caches are stored in a subdirectory (which is created before running the tests) fails, as neither `phpunit` nor `infection` create the subdirectory when running mutation tests.

Not sure if this is something that should be handled on `phpunit`'s side, or something `infection` should handle.

I think, though, that the easiest would be if this were handled by `phpunit`.

<details>
 <summary>test/Unit/phpunit.xml</summary>
 
 ```xml
<phpunit
    cacheResult="true"
    cacheResultFile="../../.phpunit/unit-result.cache"
>
</phpunit>
 ```
</details>

<details>
 <summary>Output with issue</summary>
 
 ```
[ERROR] Project tests must be in a passing state before running Infection.

         Infection runs the test suite in a RANDOM order. Make sure your tests do not have hidden dependencies.

         You can add these attributes to `phpunit.xml` to check it: <phpunit executionOrder="random"
         resolveDependencies="true" ...

         If you don't want to let Infection run tests in a random order, set the `executionOrder` to some value, for
         example <phpunit executionOrder="default"

         Check the executed command to identify the problem:
         '/Users/am/Sites/localheinz/composer-json-normalizer/vendor/phpunit/phpunit/phpunit' '--configuration'
         '/var/folders/yf/j36qt8n126sf70t0rr_mms500000gn/T/infection/phpunitConfiguration.initial.infection.xml'

         PHPUnit reported an exit code of 143.

         Refer to the PHPUnit's output below:

         STDOUT:

         PHPUnit 7.4.3 by Sebastian Bergmann and contributors.

         Runtime:       PHP 7.1.22 with Xdebug 2.6.1
         Random seed:   1542452711
         Configuration:
         /private/var/folders/yf/j36qt8n126sf70t0rr_mms500000gn/T/infection/phpunitConfiguration.initial.infection.xml

         ...............................................................  63 / 346 ( 18%)
         ............................................................... 126 / 346 ( 36%)
         ............................................................... 189 / 346 ( 54%)
         ............................................................... 252 / 346 ( 72%)
         ............................................................... 315 / 346 ( 91%)
         ...............................                                 346 / 346 (100%)

         STDERR:

         PHP Warning:
         file_put_contents(/private/var/folders/yf/j36qt8n126sf70t0rr_mms500000gn/T/infection/../../.phpunit/unit-result
         .cache): failed to open stream: No such file or directory in
         /Users/am/Sites/localheinz/composer-json-normalizer/vendor/phpunit/phpunit/src/Util/TestResultCache.php on line
         82
         PHP Stack trace:


 ```
</details>

**Comment by localheinz** [comment]
Fixed with https://github.com/sebastianbergmann/phpunit/commit/4dec3dd406f7c0183831e0ed8ebf30cc2260d5e1.

**Comment by borNfreee** [comment]
Thank you @localheinz :)

**Comment by sanmai** [comment]
Is it correct that Infection still failing even after that fix? Sure, with a more verbose message, but still.

Also, what if someone runs Infection under a root user, or just `$TMPDIR/infection/../../.phpunit/` happens to be writable? Will it be polluted with random files?

IMO this should be handled by Infection too, just like #570

**Comment by localheinz** [comment]
@borNfreee @sanmai 

I have just updated `phpunit/phpunit` to `7.4.5.` in `localheinz/phpstan-rules` and ran into this issue in CI. Locally it works fine, but in CI it fails with

```
[ERROR] Project tests must be in a passing state before running Infection.     
                                                                                
         Infection runs the test suite in a RANDOM order. Make sure your tests  
         do not have hidden dependencies.                                       
                                                                                
         You can add these attributes to `phpunit.xml` to check it: <phpunit    
         executionOrder="random" resolveDependencies="true" ...                 
                                                                                
         If you don't want to let Infection run tests in a random order, set the
         `executionOrder` to some value, for example <phpunit                   
         executionOrder="default"                                               
                                                                                
         Check the executed command to identify the problem:                    
         '/home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/php
         unit' '--configuration'                                                
         '/tmp/infection/phpunitConfiguration.initial.infection.xml'            
                                                                                
         PHPUnit reported an exit code of 143.                                  
                                                                                
         Refer to the PHPUnit's output below:                                   
                                                                                
         STDOUT:                                                                
                                                                                
         PHPUnit 7.4.5 by Sebastian Bergmann and contributors.                  
                                                                                
         Runtime:       PHP 7.2.12 with Xdebug 2.6.1                            
         Random seed:   1543962983                                              
         Configuration:                                                         
         /tmp/infection/phpunitConfiguration.initial.infection.xml              
                                                                                
         ...............................................................  63 /  
         109 ( 57%)                                                             
         ..............................................                  109 /  
         109 (100%)                                                             
         Fatal error: Uncaught PHPUnit\Runner\Exception: Cannot create directory
         "/tmp/infection/../../.phpunit/integration.cache" for result cache file
         in                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php on line 81                                    
                                                                                
         PHPUnit\Runner\Exception: Cannot create directory                      
         "/tmp/infection/../../.phpunit/integration.cache" for result cache file
         in                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php on line 81                                    
                                                                                
         Call Stack:                                                            
             0.0002     418328   1. {main}()                                    
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/phpu
         nit:0                                                                  
             0.0039     950272   2. PHPUnit\TextUI\Command::main()              
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/phpu
         nit:61                                                                 
             0.0039     959064   3. PHPUnit\TextUI\Command->run()               
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         TextUI/Command.php:162                                                 
             0.0430    3947880   4. PHPUnit\TextUI\TestRunner->doRun()          
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         TextUI/Command.php:206                                                 
             4.8612   18349248   5.                                             
         PHPUnit\Runner\ResultCacheExtension->executeAfterLastTest()            
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         TextUI/TestRunner.php:626                                              
             4.8612   18349248   6. PHPUnit\Runner\ResultCacheExtension->flush()
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Runner/ResultCacheExtension.php:86                                     
             4.8612   18349248   7. PHPUnit\Runner\TestResultCache->persist()   
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Runner/ResultCacheExtension.php:26                                     
             4.8612   18349248   8. PHPUnit\Runner\TestResultCache->saveToFile()
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php:71                                            
                                                                                
                                                                                
                                                                                
         STDERR:                                                                
                                                                                
         PHP Fatal error:  Uncaught PHPUnit\Runner\Exception: Cannot create     
         directory "/tmp/infection/../../.phpunit/integration.cache" for result 
         cache file in                                                          
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php:81                                            
         Stack trace:                                                           
         #0                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php(71):                                          
         PHPUnit\Runner\TestResultCache->saveToFile()                           
         #1                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Runner/ResultCacheExtension.php(26):                                   
         PHPUnit\Runner\TestResultCache->persist()                              
         #2                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Runner/ResultCacheExtension.php(86):                                   
         PHPUnit\Runner\ResultCacheExtension->flush()                           
         #3                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         TextUI/TestRunner.php(626):                                            
         PHPUnit\Runner\ResultCacheExtension->executeAfterLastTest()            
         #4                                                                     
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         TextUI/Command.php(206):                                               
         PHPUnit\TextUI\TestRunner->doRun(Object(PHPUnit\Framework\T in         
         /home/travis/build/localheinz/phpstan-rules/vendor/phpunit/phpunit/src/
         Util/TestResultCache.php on line 81                                    
                                                                                
The command "vendor/bin/infection --ignore-msi-with-no-mutations --min-covered-msi=96 --min-msi=96" exited with 1.
```

For reference, see

* https://travis-ci.com/localheinz/phpstan-rules/jobs/162844743#L695-L802
* https://travis-ci.com/localheinz/phpstan-rules/jobs/162844957#L696-L803
* https://github.com/localheinz/phpstan-rules/pull/50

**Comment by sanmai** [comment]
Definitely these options have to be removed from the configuration.

**Comment by localheinz** [comment]
Fixed with #582.

**Comment by matiux** [comment]
Could anyone tell me if my problem falls into this case? https://github.com/infection/infection/issues/1706
I've been struggling for days
